### PR TITLE
Aggregate signature validation

### LINKED ISSUE #243: Fix `validate_attestation_aggregate_signature`
### What is wrong?

https://github.com/ethereum/eth2.0-specs/pull/532/

### How can it be fixed

- Update `validate_attestation_aggregate_signature` in `eth2/beacon/state_machines/forks/serenity/block_validation.py`
- Plus: go through again validate_attestation logic, compare the with the spec to confirm that we didn't miss other new spec changes.

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

Fixes #243.

### How was it fixed?

Updated the logic for validating an aggregate signature on an attestation.

Note: there is some refactoring here to expose various bits of logic to avoid duplication either in code or computation.

Summary of changes:

- Added helper `get_crosslink_committee_for_attestation`
    - Returns the given crosslink committee for the given attestation
    - Raises ValidationError if there is no such shard
    - Covered by existing test for `get_attestation_participants`
- Added helper `get_members_from_bitfield`
    - Returns the val_index from the committee if they are marked in the bitfield
    - Should not return anyone not marked in the bitfield
    - Covered by existing test for `get_attestation_participants`
- Added a helper `generate_aggregate_pubkeys_from_indices`
    - Returns the aggregate pubkey for a set of validators indicated by index
    - Tested by existing tests for `generate_aggregate_pubkeys`
- Updates `validate_attestation_aggregate_signature`
    - Ensure custody_bitfield is empty
    - Ensures aggregation_bitfield is not empty
    - Ensures that is a bit is not set in aggregate_bitfield then it is also not set in the custody_bitfield
    - Uses bls.verify_multiple to verify both sets of participants, partitioned by custody bit
- Ensure we can only process up to MAX_ATTESTATIONS in a given block
- added tests for the new bits of functionality listed above not covered by existing tests

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by ralexstokes** [comment]
i need to update tests that were broken in any refactoring and add new tests to cover the new validation logic so [WIP] for now. opened this PR in case anyone wants to get started reviewing ahead of time

**Comment by ralexstokes** [comment]
moving from [WIP] to ready for review :) 

@hwwhww @djrtwo .... anyone else?

the only caveat i would add is that we can get more granular with testing the ins-and-outs of the BLS verification itself which at the moment is relying on correctness by proxy of the same sorts of tests with `SlashableAttestation` passing... i think we should prefer breadth (filling in more of the missing spec) over depth at this point but i'm happy to consider adding more granular tests if someone thinks it is important.

**Comment by hwwhww** [comment]
> the only caveat i would add is that we can get more granular with testing the ins-and-outs of the BLS verification itself which at the moment is relying on correctness by proxy of the same sorts of tests with `SlashableAttestation` passing... i think we should prefer breadth (filling in more of the missing spec) over depth at this point but i'm happy to consider adding more granular tests if someone thinks it is important.


Do you mean the tests of BLS module? (`eth2.beacon._utils.bls`) or the upper layer functions which use BLS?
For the BLS module itself, it seems our current strategy is generating more testcase in https://github.com/ethereum/eth2.0-test-generators, and depends on other clients to verify the correctness of common test vectors together.

File: eth2/beacon/committee_helpers.py [hunk 1]
```diff
@@ -232,27 +232,21 @@ def get_beacon_proposer_index(state: 'BeaconState',
...
-                                 shard_count: int) -> Iterable[ValidatorIndex]:
+def get_crosslink_committee_for_attestation(state: 'BeaconState',
+                                            attestation_data: 'AttestationData',
+                                            genesis_epoch: EpochNumber,
+                                            epoch_length: int,
+                                            target_committee_size: int,
+                                            shard_count: int) -> Iterable[ValidatorIndex]:
     """
-    Return the participant indices at for the ``attestation_data`` and ``bitfield``.
+    Returns the specific crosslink committee concerning the given ``attestation_data``.
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/committee_helpers.py]: good catch

File: eth2/beacon/helpers.py [hunk 2]
```diff
@@ -215,19 +215,25 @@ def get_pubkey_for_indices(validators: Sequence['ValidatorRecord'],
...
+    return map(
+        bls.aggregate_pubkeys,
+        map(get_pubkeys, indices),
+    )
+
+
 def generate_aggregate_pubkeys(
         validators: Sequence['ValidatorRecord'],
-        slashable_attestation: 'SlashableAttestation') -> Iterable[BLSPubkey]:
+        slashable_attestation: 'SlashableAttestation') -> Sequence[BLSPubkey]:
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/helpers.py]: will do -- `Tuple` is more concrete than `Sequence` so this couples the function signature to another function (which may change) but I don't think it is a big deal here

File: eth2/beacon/state_machines/forks/serenity/block_validation.py [hunk 3]
```diff
@@ -276,6 +282,52 @@ def validate_attestation_shard_block_root(attestation_data: AttestationData) ->
...
+    # NOTE: to be removed in phase 1.
+    empty_custody_bitfield = b'\x00' * len(attestation.custody_bitfield)
+    if attestation.custody_bitfield != empty_custody_bitfield:
+        raise ValidationError(
+            "Attestation custody bitfield is not empty.\n"
+            "\tFound: %s, Expected %s" %
+            (
+                attestation.custody_bitfield,
+                empty_custody_bitfield,
+            )
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/block_validation.py]: gotcha -- yeah i was mirroring some other code for this style but personally prefer the `f-strings` so i will update

**Comment by ralexstokes** [comment]
> Do you mean the tests of BLS module? (`eth2.beacon._utils.bls`) or the upper layer functions which use BLS?

I mean specifically this block of code: https://github.com/ralexstokes/trinity/blob/12394c47be55b44fb14d1b7690ba44ef7d9c4d42/eth2/beacon/state_machines/forks/serenity/block_validation.py#L364-L391

There is a test for the high-level behavior if the signature is incorrect, but we could add tests for essentially each line of behavior in that block -- I would say that is getting pretty granular, however, which is why I wanted to get another opinion. If the existing PR seems fine, then we are good to go :)

File: eth2/beacon/state_machines/forks/serenity/block_validation.py [hunk 4]
```diff
@@ -276,6 +282,52 @@ def validate_attestation_shard_block_root(attestation_data: AttestationData) ->
...
+        )
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/block_validation.py]: yep, good catch; i will consolidate as much as makes sense

**Comment by ralexstokes** [comment]
@hwwhww incorporated all of the above feedback :)

**Comment by hwwhww** [comment]
@ralexstokes  I see. Hmm, I think indeed breadth and depth are both important; and speaking of depth, I think we should put more efforts in challenging/polishing the spec than our implementations for research-y benefit. Although, the lesson from history, as the spec is not as stable as some finished EIP, many features have been overhauled, even though it's much more stable than last year; therefore, spending large amounts of in *more* granular function tests at this point might be lower ROI than coverage than finishing the spec-syncing.

I like how you made one function into multiple steps for getting neat and easier to test. 👍👍👍 *If* you find that some refactoring + improving tests need more time than the ticket itself, and it's hard to evaluate it, maybe create a new ticket in the backlog?

/cc @pipermerriam @djrtwo

File: tests/eth2/beacon/state_machines/test_attestation_validation.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 from eth.constants import (
```
Comment (hwwhww) [inline tests/eth2/beacon/state_machines/test_attestation_validation.py]: It seems we should move `tests/eth2/beacon/state_machines/test_attestation_validation.py` to `tests/eth2/beacon/state_machines/forks/test_serenity_block_attestation_validation.py` for symmetry.

**Comment by djrtwo** [comment]
@hwwhww Generally agreed. It's a balance, but we need to charge ahead on having the state transition function complete because the expectation is that py-evm/trinity will generate the state tests.

**Comment by ralexstokes** [comment]
@hwwhww addressed the test rename above and rebased off master -- should be ready for merge if it looks good to you :)

**Pull request merged**